backend/drm/drm.c

```diff
@@ -1640,3 +1649,101 @@ void destroy_drm_connector(struct wlr_drm_connector *conn) {
 	wl_list_remove(&conn->link);
 	free(conn);
 }
+
+/* TODO: make the function return a `wlr_drm_lease` to provide a destroy event
+ * that can be fired when the kernel notifies us through uevent that the lease
+ * has been destroyed
+ */
+int wlr_drm_create_lease(struct wlr_output **outputs, size_t n_outputs,
+		uint32_t *lessee_id) {
+	assert(outputs);
+
+	if (n_outputs == 0) {
+		wlr_log(WLR_ERROR, "Can't lease 0 outputs");
+		return -1;
+	}
+
+	struct wlr_drm_backend *drm =
+			get_drm_backend_from_backend(outputs[0]->backend);
+
+	int n_objects = 0;
+	uint32_t objects[4 * n_outputs + 1];
+	for (size_t i = 0; i < n_outputs; ++i) {
+		struct wlr_drm_connector *conn =
+				get_drm_connector_from_output(outputs[i]);
+		assert(conn->lessee_id == 0);
+
+		if (conn->backend != drm) {
+			wlr_log(WLR_ERROR, "Can't lease output from different backends");
+			return -1;
+		}
+
+		objects[n_objects++] = conn->id;
+		wlr_log(WLR_DEBUG, "Connector %d", conn->id);
+
+		if (!conn->crtc) {
+			wlr_log(WLR_ERROR, "Connector has no CRTC");
+			return -1;
+		}
+
+		objects[n_objects++] = conn->crtc->id;
+		wlr_log(WLR_DEBUG, "CRTC %d", conn->crtc->id);
+
+		objects[n_objects++] = conn->crtc->primary->id;
+		wlr_log(WLR_DEBUG, "Primary plane %d", conn->crtc->primary->id);
+
+		if (conn->crtc->cursor) {
+			wlr_log(WLR_DEBUG, "Cursor plane %d", conn->crtc->cursor->id);
+			objects[n_objects++] = conn->crtc->cursor->id;
+		}
+	}
+
+	assert(n_objects != 0);
+
+	wlr_log(WLR_DEBUG, "Issuing DRM lease with the %d objects", n_objects);
+	int lease_fd = drmModeCreateLease(drm->fd, objects, n_objects, 0,
+			lessee_id);
+	if (lease_fd < 0) {
+		return lease_fd;
+	}
+
+	wlr_log(WLR_DEBUG, "Issued DRM lease %"PRIu32, *lessee_id);
+	for (size_t i = 0; i < n_outputs; ++i) {
+		struct wlr_drm_connector *conn =
+				get_drm_connector_from_output(outputs[i]);
+		conn->lessee_id = *lessee_id;
+		conn->crtc->lessee_id = *lessee_id;
```

Comment:
Don't we need to destroy the `wlr_output` somewhere here, to make sure the compositor doesn't see the connector anymore?

Reply:
There was a call to set the connector as "disconnected" but I experienced a lot of issue with the connector being recreated each time a lease is revoked. Now the output is just given to the wlr_drm_lease_v1_manager once and it is properly destroyed when the display is powered down.

Reply:
That doesn't sound like a good idea to me. The compositor will see a connected output and might try to use it.

Reply:
It should not, the non-desktop property is filled and the compositor should ignore it.

Reply:
Compositors might want to display some kind of pattern on non-desktop outputs. Compositors might lease desktop outputs.

Reply:
The `wlr_output` are needed by the lease manager, what would be a good approach?

Reply:
Return a `struct wlr_drm_lease_connector_v1 *` when offering the output, and allow the compositor to destroy it to withdraw the output and get back a `wlr_output`.

But alright, let's just go with the current approach for now. This is an unstable interface, we can improve it later.